### ISSUE TITLE
Support for vector splats (scalar operations on vectors)

### ISSUE THREAD

**Issue by jespertheend** [gpuweb/gpuweb#1520]
Coming from https://github.com/gpuweb/gpuweb/issues/1518#issuecomment-799453939
I think it'd be nice to be able to do
```
pos = pos / pos.w;
```
rather than
```
const w : f32 = pos.w;
pos = pos / vec4<f32>(w,w,w,w);
```

**Comment by dneto0** [comment]
I agree this is commonly supported by mainstream shader languages.
The scalar is automatically converted to a vector where:
- the number of elements corresponds to the other operand of the arithmetic operator
- each component is a copy of the scalar.

For some reason this is called a "splat".  (I saw you got this advice already, but I wanted to establish common terminology for this issue.)

**Comment by kvark** [comment]
IIRC, SPIR-V doesn't support this. Are we certain this needs to be supported in WGSL MVP?
It seems neat to me that this would nudge people towards writing `pos * (1.0 / w)`, which is more efficient (but obviously not always the same).

**Comment by dneto0** [comment]
I'm certain it's not _necessary_ for MVP. (The proof is that the user can write everything they need, by using a vector-type constructor with repeated elements.)

We have a design guideline that WGSL has no automatic conversions between types.  But we could add vector constructors that take a scalar. This would  add 3 more rows to the "Vector constructor type rules" table at https://gpuweb.github.io/gpuweb/wgsl.html#type-constructor-expr, one for each vector size.

So:
       var pos : vec4<f32> = ...
       const w: f32 = pos.w;  // just to pin down that it's an f32
       // splats the scalar into a vector of the right size.  But you still need to know the vector size.
       pos = pos / vec4<f32>(w);

**Comment by jespertheend** [comment]
> For some reason this is called a "splat". (I saw you got this advice already, but I wanted to establish common terminology for this issue.)

I was hesitant to use this terminology because I couldn't find anything about this anywhere on the internet. I'll change the title though.

**Comment by dneto0** [comment]
This was discussed at the end of the [2021-03-16 meeting](https://docs.google.com/document/d/18mbpgi_-OirqNXBPmi9dOAdsQd5en-80v5kmZvAUC5U/edit#)

There was support for vector constructors that take a scalar argument, as I described above.

I'll make a PR for that.

It was a very quick discussion, so I hesitate to mark this as "wgsl-resolved"

**Comment by jdashg** [comment]
Not quite fixed as-requested yet.

**Comment by kvark** [comment]
Do we really need to do anything here? I thought we'd just close it and re-open if the current solution is not enough.

**Comment by 11mhg** [comment]
+1 for vector splats and other similar broadcast operators. They're great for lowering mental load when you're doing tons of math.

**Comment by dneto0** [comment]
Discussed in meeting of [2021-03-23](https://docs.google.com/document/d/19inX49uQBMkWx1lSuPkFWJ8fGRJUGMNXmkUj5TJWnwY/edit#heading=h.o1a3xkmhqrws)

* Some desire to support some mixed-vector-scalar operations for basic arithmetic: + - /     (multiply already supported)
* Some opposition to supporting that for anything other than multiply
* TODO: @jdashg to survey other common shader languages to see what they support
* TODO: @pow2clk to  see what HLSL supports

**Comment by 11mhg** [comment]
This is well defined in the GLSL spec and even extends to Matrices. [The GLSL 4.6 spec ](https://www.khronos.org/registry/OpenGL/specs/gl/GLSLangSpec.4.60.pdf) says that 

> The  arithmetic  binary  operators  add  (+),  subtract  (-),  multiply  (*),  and  divide  (/)  operate  on integer  and  floating-point  scalars,  vectors,  and  matrices.  If  the  fundamental  types  in  the operands do not match, then the conversions from “Implicit Conversions” are applied to create matching  types.  All  arithmetic  binary  operators  result  in  the  same  fundamental  type  (signed integer,  unsigned  integer,  single-precision  floating-point,  or  double-precision  floating-point)  as the  operands  they  operate  on,  after  operand  type  conversion.  After  conversion,  the  following cases are valid

and follows on to say

> One operand is a scalar, and the other is a vector or matrix. In this case, the scalar operation is  applied  independently  to  each  component  of  the  vector  or  matrix,  resulting  in  the  same size vector or matrix.

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-03-23-Minutes#support-for-vector-splats-scalar-operations-on-vectors-1520">WGSL meeting minutes 2021-03-23</a></summary>

*   E.g. `vec(x,y) * 2.0`
*   DN wrote PR [1527](https://github.com/gpuweb/gpuweb/pull/1527) for vector-constructor-from-scalar;  DM landed it.
*   Deliberately does not address the naked scalar case as originally requested
*   MM: Why is multiply privileged
*   DN: linear algebra building block.
*   BC: As a graphics programmer, I want to divide a vector by a scalar.
*   JG: `gl_Position = vec4(pos01 * 2.0 - 1.0, 0.0, 1.0);`
*   MM: Thought we had a rule about no implicit cast.
    *   Here’s a counter proposal: Don’t allow “vector op scalar” and “matrix op scalar” for any operation. Instead, expose constructors / standard library functions to turn those into “vector op vector” and “matrix op matrix”
*   [...]
*   MM: If someone made the popularity argument, I would be convinced by that. The argument being “if every popular shading language supports ‘vector * scalar’ but not ‘vector / scalar’ then that’s a pretty strong argument that we should follow”
*   DM: Maybe it’s better to force people to be really explicit?
*   DN: Every new feature costs test time.
*   JG: I’ll survey other shader languages to see what they have.
*   GR: I volunteer to get the answer for HLSL.
*   GR (later): HLSL accepts every arithmetic operation between vectors and scalars and matrices and vectors for both DXC and FXC

</details>

**Comment by dneto0** [comment]
The implicit conversions listed in the GLSL 4.6 spec are the same as the 
- the "usual arithmetic conversions" in C:
     - signed int to unsigned int of same width, or float
     - unsigned into to float
     - float to double
- plus also corresponding vector-to-vector cases (where the component types are converted as above)

It's the second clause you quoted that does the widening.

**Comment by dneto0** [comment]
Does this interact with a future where we can write `1` instead of `1u` or `1.0`?

**Comment by dneto0** [comment]
the GLSL spec says:
- for + - and /  matrix cases, the matices must be same dimensions and the operation is done component-wise, after implicit conversions.
- for * and matrix cases it is matrix multiply. (I hope that's not a foot-gun)

**Comment by dneto0** [comment]
Discussed in meeting of [2021-03-29](https://docs.google.com/document/d/173UCbLQC0kUESMaZ-1dB8jZQdglRWSGy98pJXWdakHE/edit#heading=h.o1a3xkmhqrws)

Consensus appears to be: support the intersection of HLSL,GLSL,MSL.  @litherum  to write down what that is.

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-03-29-(Monday!)-Minutes#support-for-vector-splats-scalar-operations-on-vectors-1520">WGSL meeting minutes 2021-03-29</a></summary>

*   E.g. `vec(x,y) * 2.0`
*   Previously:
*   JG: I’ll survey other shader languages to see what they have.
*   GR: I volunteer to get the answer for HLSL.
*   GR (later): HLSL accepts every arithmetic operation between vectors and scalars and matrices and ~~vectors~~ scalars for both DXC and FXC
*   GR: HLSL supports bitwise ops too in all scalar with sequential types
*   GR: and also the transposition of any and all operands
    *   MM: Metal supports vector {*,/,+,-} scalar
    *   MM: Metal supports scalar {*,/,+,-} vector
    *   MM: Metal supports matrix {*,/} scalar
    *   MM: Metal supports scalar * matrix
    *   MM: Metal supports matrix * vector
    *   MM: Metal supports vector * matrix
    *   DN: on the issue, Mohammed found what GLSL supports.  [https://github.com/gpuweb/gpuweb/issues/1520#issuecomment-806004434](https://github.com/gpuweb/gpuweb/issues/1520#issuecomment-806004434) Pretty liberal mixing of scalar-vector and scalar-matrix, and  +,-,/ for matrix-matrix of same dimensions.
    *   AB: GLSL also supports some bit-operations over these.
    *   JG: I mainly care about arithmetic operations.
    *   MM: Propose WGSL support the intersection of GLSL,HLSL,MSL
    *   DM: When do we stop. What about shifts?
    *   MM: Think the result falls out. See GLSL restrictions.
    *   AB: GLSL supports i
    *   DN: Small concern: Does this block us from having e.g. `1` being contextually 1i or 1.0f?
    *   AB: GLSL does the conversion before splatting
    *   MM: I think we have solutions here
    *   GR: Do we want to include mat*scalar here?
    *   MM: Probably, more for mult than div.
    *   JG: Ok, I will write up proposal based on intersections
</details>

**Comment by kvark** [comment]
@dneto0 what about `mix`?
In SPIR-V, it's expected to work with vectors, but in GLSL it can accept a float for the last parameter, while others are vectors.

**Comment by dneto0** [comment]
> @dneto0 what about mix?

See FMix in the [GLSL.std.450 extended instruction set](https://www.khronos.org/registry/spir-v/specs/1.0/GLSL.std.450.html):  It's all operands-and-result f32 or all operands-and-result vector-of-f32.   So if you wanted WGSL to support scalar in last param, you'd have to polyfill for SPIR-V.

**Comment by dneto0** [comment]
Discussed in [meeting of 2021-03-30](https://docs.google.com/document/d/1gsji6SmK50HhCeK6H6tawPXaeUWWrhLRZbMiRpbAm5U/edit#heading=h.o1a3xkmhqrws)

* The list of variations is shrinking, e.g. no matrix + scalar or scalar+matrix.
* @jdashg to re-state the intended supported set of operations.  @dneto0 to write up for the spec.

**Comment by jdashg** [comment]
<details><summary><a href="https://github.com/gpuweb/gpuweb/wiki/WGSL-2021-03-30-Minutes#support-for-vector-splats-scalar-operations-on-vectors-1520">WGSL meeting minutes 2021-03-30</a></summary>

*   E.g. `vec(x,y) * 2.0`
*   Previously:
    *   JG: Ok, I will write up proposal based on intersections
*   PR: [https://github.com/gpuweb/gpuweb/pull/1575](https://github.com/gpuweb/gpuweb/pull/1575) 
*   DN: should be a table instead of a list, will do that editorial change and submit a new PR. Otherwise PR is fine.
*   Followup
*   GR: Matrix divide by scalar [or was that the other way around]  Is it in the intersection, and is it an oversight.
*   JG: MSL spec doesn’t say it allows it.  MSL spec might be incomplete.
*   _Discussion_
*   DM: Should reconsider matrix + scalar. Think it’s not useful, and is ambiguous.  Might have wanted a diagonal matrix.
*   JG: Generally, it’s always component-wise. Exception is matrix * vector (and vector * matrix).
*   DM: When discussing matrix constructor of scalar, we decided that was ambiguous/confusing/not-clear. Should not reintroduce it implicitly here.
*   JG: Think we want multiply matrix by scalar. By analogy would want to allow adding.
*   MM: MSL doesn’t allow matrix + scalar or scalar + matrix. 
*   JG: Nevermind then! I’ll update the proposed list of supported operations.

</details>

**Comment by jdashg** [comment]
Closed by #1611.
